File: 백준/Silver/6603. 로또/로또.cc
#include <bits/stdc++.h>
#define endl "\n"

using namespace std;

int input;
int k = 1;

int number[13];
int result[13];

int visited[13] = { 0, };

// k개의 숫자를 뽑았을때 중복되지 않게 모든 경우의 조합을 구하라. 
// 조합 문제 - DFS

void dfs( int cnt )
{
	if (cnt == 6)
	{
		for (int i = 0; i < 6; i++)
		{
			cout << result[i+1] << " ";
		}

		cout << endl;
		return;
	}

	for (int i = cnt; i < k; i++)
	{
		if (!visited[i])
		{
			visited[i] = 1;
			if (number[i] > result[cnt])
			{
				result[cnt+1] = number[i];
				dfs(cnt + 1);
			}
			visited[i] = 0;
		}
	}
}

int main()
{
	while (1)
	{
		cin >> k;

		if (k == 0)
		{
			break;
		}

		for (int i = 0; i < k; i++)
		{
			cin >> number[i];
		}

		dfs(0);

		cout << endl;
	}
	return 0;
}

File: 백준/Silver/6603. 로또/로또.java
import java.util.*;
import java.io.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb;

    static int[] arr, result;
    static int k;
    static boolean[] visited;

    public static void main(String[] args) throws Exception {
        while(true) {

            st = new StringTokenizer(br.readLine());
            sb = new StringBuilder();
            k = Integer.parseInt(st.nextToken());

            if (k == 0) {
                break;
            }

            arr = new int[k];

            for (int i = 0 ; i < k ; i ++) {
                arr[i] = Integer.parseInt(st.nextToken());
            }

            visited = new boolean[k];
            result = new int[6];

            dfs(0, 0);

            System.out.println(sb);
        }
    }

    static void dfs(int cur, int idx) {
        if (cur == 6) {
            for (int i = 0 ; i < 6 ; i ++) {
                sb.append(result[i]).append(' ');
            }
            sb.append('\n');

            return;
        }

        for (int i = idx ; i < k ; i ++) {
            if (!visited[i]) {
                visited[i] = true;
                result[cur] = arr[i];
                dfs(cur + 1, i);
                visited[i] = false;
            }
        }
    }
}